---
title: "Benchmarking sgRNA libraries: Build one mixed library for reduced set of essentials and non-essentials"
author: "Alex Kalinka"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: hide
    depth: 3
    highlight: tango
    number_sections: true
    theme: spacelab
    toc: true
    toc_float: true
---

# Setup

```{r setup}
set.seed(1)
options(warn=-1)
suppressMessages(library(knitr))
suppressMessages(library(tidyr))
suppressMessages(library(magrittr))
suppressMessages(library(ggplot2))
suppressMessages(library(dplyr))
suppressMessages(library(combinat))
#suppressMessages(library(org.Hs.eg.db))
#suppressMessages(library(biomaRt))
#suppressMessages(library(Biostrings))


filter <- dplyr::filter
select <- dplyr::select
rename <- dplyr::rename
arrange <- dplyr::arrange

context_dep_genes <- c("TP53","MDM2","MDM4","PPM1D","KRAS","WRN","PIK3CA",
                       "PARP1","PARP2","BCL2L1","MAP2K1","MAPK1")

# hg38.
#ensembl <- useMart("ensembl",dataset="hsapiens_gene_ensembl")

# hg37 (for croatan).
#ensembl.hg37 <- useMart("ensembl",dataset="hsapiens_gene_ensembl", host="http://feb2014.archive.ensembl.org")

get_standard_symbols <- function(gene_names, type = "sym"){
  if(type=="sym"){
    syms <- AnnotationDbi::select(org.Hs.eg.db, gene_names, "SYMBOL", "ALIAS")
  }else{
    syms <- AnnotationDbi::select(org.Hs.eg.db, gene_names, "ENSEMBL", "ALIAS") %>%
      rename(SYMBOL = ENSEMBL)
  }
  syms %<>%
    rowwise() %>%
    mutate(SYMBOL = ifelse(is.na(SYMBOL),ALIAS,SYMBOL)) %>%
    ungroup() %>%
    # Resolve many:1 mappings.
    group_by(ALIAS) %>%
    mutate(SYMBOL = ifelse(n() > 1, ALIAS, SYMBOL)) %>%
    filter(!duplicated(ALIAS)) %>%
    ungroup()
  return(syms$SYMBOL)
}

add_syms <- function(data, gene_col){
  reps <- data %>%
    group_by(!!sym(gene_col)) %>%
    summarise(num = n()) %>%
    ungroup()
  syms <- get_standard_symbols(unique(unlist(data[,gene_col])))
  if(length(syms) != length(unique(unlist(data[,gene_col]))))
    stop("unequal number of symbols returned")
  data %<>%
    mutate(gene_symbol = rep(syms, reps$num))
  return(data)
}


# Functions to:
# 1. Get missing gene coding regions.
# 2. Build a FASTA file of the missing genes.
# 3. Build into a bowtie reference.
# 4. Short-read alignment of the gRNAs.
build_missing_gene_bowtie_ref <- function(genes, fasta_out, bowtie_ref_dir, hsapiens_mart){
  # Convert to Ensembl IDs.
  ens <- get_standard_symbols(genes, type = "ens")
  # 1. Download coding regions from biomart.
  seqs <- getBM(attributes = c("hgnc_symbol","coding"), 
                filters="hgnc_symbol", values=genes, mart=hsapiens_mart) %>%
    rbind(getBM(attributes = c("hgnc_symbol","coding"), 
                filters="ensembl_gene_id", values=ens[grepl("^ENS",ens)], mart=hsapiens_mart)) %>%
    filter(!is.na(coding)) %>%
    rowwise() %>%
    filter(coding != "Sequence unavailable") %>%
    mutate(len = nchar(coding)) %>%
    ungroup() %>%
    # 2. Select longest seq for each gene.
    group_by(hgnc_symbol) %>%
    filter(len == max(len)) %>%
    filter(n()==1) %>%
    ungroup()
  seq_inp <- seqs$coding
  names(seq_inp) <- seqs$hgnc_symbol
  seqs <- DNAStringSet(seq_inp)
  # 3. BUild FASTA.
  writeXStringSet(seqs, fasta_out, format = "fasta")
  # 4. Make bowtie reference.
  bowtie2_build(fasta_out, bt2Index=bowtie_ref_dir, overwrite = T)
  return(seqs)
}


bowtie2_aln <- function(seqs, ref_base, fasta_out, sam_out){
  # Write out seqs as FASTA.
  writeXStringSet(seqs, fasta_out, format = "fasta")
  # Run bowtie2.
  bwt_cmd <- paste("bowtie2 -x ",ref_base," -U ",fasta_out," -S ",sam_out," -f --no-unal",sep="")
  system(bwt_cmd, wait = T)
}

read_sam <- function(sam_out){
  # Read and process alignments.
  aln <- read.delim(sam_out, header=F, comment.char = "@", sep="\t", stringsAsFactors = F) %>%
    select(V2, V3, V6, V10) %>%
    filter(V6 == "20M" & !duplicated(V10)) %>%
    rowwise() %>%
    # reverseComplement for reverse strand mappings:
    # (use simple equality for bitwise flag since it only takes 2 values here).
    mutate(V10 = ifelse(V2 == 16,as.character(reverseComplement(DNAString(V10))),V10)) %>%
    ungroup()
  return(aln)
}

```

# Approach

1. Find gene symbols that are common to all libraries.
2. Pick all of David Walter's genes that are common across all the libraries.
3. Top up essentials to make 250 in total using Sanger pan-cancer genes.
4. Top up non-essentials to make 500 in total using Bagel non-essentials.
5. Add 12 context-dependent genes.
6. Add 100 control guides from Yusa v3.

# Input genes

```{r}
load(file="../data/ess.genes.rda")
ess.time <- read.csv("../data/essential_early_late.csv", stringsAsFactors = F)

genes <- read.csv("../data/benchmarking_genes.csv", stringsAsFactors = F) %>%
  mutate_at(vars(-Gene), .funs = function(x) ifelse(x == "y",T,F)) %>%
  arrange(Gene) %>%
  mutate(davids_list = T) %>%
  rbind(data.frame(Gene = c(ess.genes$pan_cancer_Sanger, ess.genes$bagel_nonessential, context_dep_genes),
                   stringsAsFactors = F) %>%
          mutate(davids_list = F, Essential = Gene %in% ess.genes$pan_cancer_Sanger,
                 Neutral_Hart = NA, Cluster = NA, Late_essential = NA,
                 Mid_essential = NA, Early_essential = NA, PanCancer_essential = NA,
                 Achilles_common_essentials = NA, Hart_core_essential_v2_2017 = NA)) %>%
    filter(!duplicated(Gene)) %>%
    mutate(pan_cancer = Gene %in% ess.genes$pan_cancer_Sanger,
           bagel_essential = Gene %in% ess.genes$bagel_essential,
           bagel_nonessential = Gene %in% ess.genes$bagel_nonessential,
           context_dependent = Gene %in% context_dep_genes,
           Early_essential = Gene %in% ess.time$gene[ess.time$essential_type=="Early"],
           Mid_essential = Gene %in% ess.time$gene[ess.time$essential_type=="Mid"],
           Late_essential = Gene %in% ess.time$gene[ess.time$essential_type=="Late"])

load(file="../data/ess.genes.rda")
yusa.v3 <- read.csv("../data/CRISPR_libraries/yusa_v3-suppl-table-S3.csv", stringsAsFactors = F) %>%
  left_join(read.table("../data/cleanr.tsv",header=T, stringsAsFactors = F), by = c("gRNA_ID" = "CODE"))

read_horizon <- function(file){
  ret <- read.csv(file, stringsAsFactors = F) %>%
    arrange(GeneSymbol) %>%
    add_syms(gene_col = "GeneSymbol")
  return(ret)
}

# Horizon gRNA libraries must be kept locally.
horizon <- rbind(read_horizon("~/data/az_cruk/GC-004650-E2 Edit-R crRNA POOL Human Drug Targets Lot 17162.csv"),
                 read_horizon("~/data/az_cruk/GC-004675-E2 Edit-R crRNA POOL Human Druggable Subsets Lot 17163.csv"),
                 read_horizon("~/data/az_cruk/GC-006500-E2 Edit-R crRNA POOL Human Genome  Lot 17155.csv"))

croatan.all <- read.table("../data/CRISPR_libraries/HG19_Library_For_Build.with_gene.txt",
                      header=F, stringsAsFactors = F)
croatan <- croatan.all %>%
  select(-V3,-V4,-V5) %>%
  rowwise() %>%
  mutate(gRNA_sequence.1 = unlist(strsplit(V2,"__"))[2],
         gRNA_sequence.2 = unlist(strsplit(V2,"__"))[6]) %>%
  ungroup() %>%
  select(-V2) %>%
  gather(key = "guide_number", value = "context_sequence", -V1) %>%
  rename(gene = V1) %>%
  arrange(gene, guide_number, context_sequence) %>%
  rowwise() %>%
  mutate(sequence = substr(context_sequence,5,24), PAM = substr(context_sequence,25,27)) %>%
  ungroup() %>%
  arrange(gene)

```

## Genes common to all libraries

```{r}
add_library <- function(file, gene_col, data = NULL){
  # Returns T or F for each gene's presence in 'genes'.
  if(!is.null(file)){
    lib <- read.csv(file, stringsAsFactors = F) 
  }else{
    lib <- data
  }
  lib %<>% mutate(gene = !!sym(gene_col))
  ret <- genes$Gene %in% lib$gene
  return(ret)
}

genes_common <- data.frame(gene = genes$Gene, stringsAsFactors = F) %>%
  mutate(Yusa_v3 = add_library(file=NULL, gene_col="Gene", data=yusa.v3),
         brunello = add_library("../data/CRISPR_libraries/brunello_grnas.csv","Target.Gene.Symbol"),
         toronto = add_library("../data/CRISPR_libraries/toronto_grnas.csv","GENE"),
         gecko = add_library("../data/CRISPR_libraries/gecko_grnas.csv","gene_id"),
         gattinara = add_library("../data/CRISPR_libraries/gattinara_grnas.csv","Gene.Symbol"),
         horizon = add_library(file=NULL,"GeneSymbol",data = horizon),
         croatan = add_library(file=NULL,"gene",croatan)) %>%
  mutate(common_count = rowSums(.[,2:ncol(.)]), common_all = common_count == ncol(.)-1)

#  Curate input genes.
# 1. All of David's genes that are common.
# 2. Fill to 250 essential, 500 nonessential.
genes %<>%
  mutate(common_all = genes_common$common_all[match(Gene,genes_common$gene)],
         keep = (davids_list & common_all),
         keep = ifelse(!keep,context_dependent,keep)) %>%
  filter(common_all)

num_ess_miss <- 250 - table(genes$keep,genes$Essential)[2,2]
num_noness_miss <- 500 - table(genes$keep,genes$Neutral_Hart)[2,2]

extra_ess <- genes$Gene[(!genes$keep & genes$Essential)][1:num_ess_miss]
extra_noness <- genes$Gene[(!genes$keep & !genes$Essential)][1:num_noness_miss]

genes %<>%
  mutate(keep = ifelse(!keep,Gene %in% c(extra_ess, extra_noness),keep)) %>%
  filter(keep)

```

## Sanity checks on input genes

```{r}
cat("Total number of genes:")
length(unique(genes$Gene))

cat("Number of Essential genes:")
sum(genes$Essential)

cat("Early essential:")
sum(genes$Early_essential)

cat("Mid essential:")
sum(genes$Mid_essential)

cat("Late essential:")
sum(genes$Late_essential)

cat("Number of Non-essential genes (Hart Neutral + Bagel non-essential):")
length(setdiff(union(genes$Gene[genes$Neutral_Hart], genes$Gene[genes$bagel_nonessential]),NA))

cat("Number of context-dependent genes:")
sum(genes$context_dependent)

cat("Pan-cancer sanity check (all counts expected on the diagonal):")
kable(table(genes$PanCancer_essential,genes$pan_cancer))

```

# Benchmarking library

## Build

```{r}
# Libraries.
yusa <- yusa.v3 %>%
  filter(!grepl("Control",gRNA_ID)) %>%
  arrange(Gene,gRNA_ID) %>%
  filter(Gene %in% genes$Gene) %>%
  mutate(library = "Yusa_v3", target_exon = NA) %>%
  rename(grna_id = gRNA_ID, gene = Gene, strand = STRAND, sequence = Guide_sequence) %>%
  arrange(gene, grna_id) %>%
  select(library, gene, grna_id, strand, target_exon, sequence)

brunello <- read.csv("../data/CRISPR_libraries/brunello_grnas.csv", stringsAsFactors = F) %>%
  arrange(Target.Gene.Symbol) %>%
  filter(Target.Gene.Symbol %in% genes$Gene) %>%
  mutate(library = "Brunello", num = 1:nrow(.), grna_id = paste("brunello.",num,sep=""),
         Strand = ifelse(Strand == "sense","+","-")) %>%
  rename(gene = Target.Gene.Symbol,
         sequence = sgRNA.Target.Sequence, strand = Strand, target_exon = Exon.Number) %>%
  arrange(gene, grna_id) %>%
  select(library, gene, grna_id, strand, target_exon, sequence)

toronto <- read.csv("../data/CRISPR_libraries/toronto_grnas.csv", stringsAsFactors = F) %>%
  arrange(GENE) %>%
  filter(GENE %in% genes$Gene) %>%
  rename(grna_id = GUIDE_ID, gene = GENE, sequence = SEQUENCE, target_exon = TARGET.EXON) %>%
  mutate(library = "Toronto",target_exon = as.integer(gsub("^exon_(\\d+)$","\\1",target_exon))) %>%
  rowwise() %>%
  mutate(strand = tail(unlist(strsplit(grna_id,"_")),1)) %>%
  ungroup() %>%
  arrange(gene, grna_id) %>%
  select(library, gene, grna_id, strand, target_exon, sequence)

gecko <- read.csv("../data/CRISPR_libraries/gecko_grnas.csv", stringsAsFactors = F) %>%
  arrange(gene_id) %>%
  filter(gene_id %in% genes$Gene) %>%
  rename(grna_id = UID, gene = gene_id, sequence = seq) %>%
  mutate(library = "Gecko", target_exon = NA, strand = NA) %>%
  arrange(gene, grna_id) %>%
  select(library, gene, grna_id, strand, target_exon, sequence)

gattinara <- read.csv("../data/CRISPR_libraries/gattinara_grnas.csv", stringsAsFactors = F) %>%
  arrange(Gene.Symbol) %>%
  filter(Gene.Symbol %in% genes$Gene) %>%
  mutate(library = "Gattinara", target_exon = NA, num = 1:nrow(.),
         grna_id = paste("gattinara.",num,sep=""), target_exon = NA, strand = NA) %>%
  rename(gene = Gene.Symbol, sequence = Barcode.Sequence) %>%
  arrange(gene, grna_id) %>%
  select(library, gene, grna_id, strand, target_exon, sequence)

horizon %<>%
  arrange(GeneSymbol) %>%
  filter(GeneSymbol %in% genes$Gene) %>%
  mutate(library = "Horizon", target_exon = NA, strand = substr(GenomicLocation,6,6)) %>%
  rename(gene = GeneSymbol, grna_id = CatalogNumber, sequence = TargetSequence) %>%
  arrange(gene, grna_id) %>%
  select(library, gene, grna_id, strand, target_exon, sequence)

croatan %<>%
  filter(gene %in% genes$Gene) %>%
  mutate(library = "Croatan", target_exon = NA, strand = NA,
         num = 1:nrow(.), grna_id = paste("Croatan.",num,sep="")) %>%
  arrange(gene, grna_id) %>%
  select(library, gene, grna_id, strand, target_exon, sequence)

```


## Final

```{r}
bm_library <- rbind(yusa, brunello, toronto, gecko, gattinara, horizon, croatan) %>%
  # Overlapping guides.
  # Must group by 19bp guides so we can include Yusa.
  rowwise() %>%
  mutate(seq_19 = ifelse(library!="Yusa_v3",substr(sequence,2,20),sequence)) %>%
  ungroup() %>%
  group_by(seq_19) %>%
  mutate(library = ifelse(n() == 1, library, paste(sort(unique(library)),collapse=".")),
         # Only keep one entry when there are duplicates.
         keep = ifelse(n() > 1, c(T,rep(F,n()-1)),T)) %>%
  ungroup() %>%
  filter(keep & !duplicated(sequence)) %>%
  select(-seq_19, -keep)

```

# Library stats

```{r}
# To produce a heatmap of gRNA sharing across libraries.
count_pairwise_lib_grnas <- function(data){
  libs <- unique(unlist(sapply(data$library, function(x) unlist(strsplit(x,"\\.")))))
  df <- as.data.frame(t(combn(libs, 2)), stringsAsFactors=F)
  df2 <- data.frame(V1=df$V2, V2=df$V1, stringsAsFactors = F)
  lc <- df %>%
    rbind(df2) %>%
    rbind(data.frame(V1=libs,V2=libs, stringsAsFactors = F)) %>%
    rename(Library_1 = V1, Library_2 = V2) %>%
    rowwise() %>%
    mutate(gRNA_count = sum((grepl(Library_1, data$library) == T & grepl(Library_2, data$library) == T))) %>%
    ungroup()
  return(lc)
}

# Percentage of gRNAs shared across libraries.
pct_shared_grnas <- function(data){
  ret <- data %>%
    group_by(Library_1) %>%
    mutate(gRNA_percent = 100*gRNA_count/sum(gRNA_count)) %>%
    ungroup()
  return(ret)
}
```

```{r}
cat("Total number of genes:")
length(unique(bm_library$gene))

cat("All genes in original set:")
all(unique(bm_library$gene) %in% genes$Gene)

cat("Total genes per library:")
kable(t(
  bm_library %>%
  summarise(Num_genes.Yusa_v3 = length(unique(gene[grepl("Yusa_v3",library)])),
            Num_genes.brunello = length(unique(gene[grepl("Brunello",library)])),
            Num_genes.toronto = length(unique(gene[grepl("Toronto",library)])),
            Num_genes.gecko = length(unique(gene[grepl("Gecko",library)])),
            Num_genes.gattinara = length(unique(gene[grepl("Gattinara",library)])),
            Num_genes.horizon = length(unique(gene[grepl("Horizon",library)])),
            Num_genes.Croatan = length(unique(gene[grepl("Croatan",library)])))
))


cat("Full overlapping gRNA library breakdown:")
kable(bm_library %>%
  group_by(library) %>%
  summarise(Num_genes = length(unique(gene))) %>%
  ungroup() %>%
  arrange(desc(Num_genes))
)

bm_lib.hmap <- bm_library %>%
  select(library) %>%
  count_pairwise_lib_grnas()

bm_lib.pct <- pct_shared_grnas(bm_lib.hmap)

ggplot(bm_lib.hmap, aes(Library_1, Library_2, fill = gRNA_count)) +
  geom_tile(color="white") +
  scale_fill_gradient(low = "palegoldenrod", high = "red") +
  ggtitle("gRNA sharing across libraries")

ggplot(bm_lib.pct, aes(Library_1, gRNA_percent)) +
  geom_bar(stat="identity") +
  facet_grid(~Library_2) +
  theme(axis.text.x = element_text(angle = -30, hjust=0, size=7)) +
  ggtitle("Percent of shared gRNAs across libraries")

```

# Add 100 Control guides from Yusa v3

```{r}


```

# Save

```{r}

```

