---
title: "Benchmarking sgRNA libraries: Build one mixed library for reduced set of essentials and non-essentials"
author: "Alex Kalinka"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: hide
    depth: 3
    highlight: tango
    number_sections: true
    theme: spacelab
    toc: true
    toc_float: true
params:
  early: !r 86
  mid: !r 77
  late: !r 87
---

# Setup

```{r setup}
options(warn=-1)
suppressMessages(library(knitr))
suppressMessages(library(tidyr))
suppressMessages(library(magrittr))
suppressMessages(library(ggplot2))
suppressMessages(library(dplyr))
# Try to convert all genes into standard gene symbols.
suppressMessages(library(org.Hs.eg.db))

filter <- dplyr::filter
select <- dplyr::select
rename <- dplyr::rename
arrange <- dplyr::arrange


get_standard_symbols <- function(gene_names){
  syms <- AnnotationDbi::select(org.Hs.eg.db, gene_names, "SYMBOL", "ALIAS") %>%
    rowwise() %>%
    mutate(SYMBOL = ifelse(is.na(SYMBOL),ALIAS,SYMBOL)) %>%
    ungroup() %>%
    # Resolve 1:many mapping.
    group_by(ALIAS) %>%
    mutate(SYMBOL = ifelse(n() > 1, ALIAS, SYMBOL)) %>%
    filter(!duplicated(ALIAS)) %>%
    ungroup()
  return(syms$SYMBOL)
}

add_syms <- function(data, gene_col){
  reps <- data %>%
    group_by(!!sym(gene_col)) %>%
    summarise(num = n()) %>%
    ungroup()
  syms <- get_standard_symbols(unique(data[,gene_col]))
  if(length(syms) != length(unique(data[,gene_col])))
    stop("unequal number of symbols returned")
  data %<>%
    mutate(gene_symbol = rep(syms, reps$num))
  return(data)
}

genes <- read.csv("../data/benchmarking_genes.csv", stringsAsFactors = F) %>%
  mutate_at(vars(-Gene), .funs = function(x) ifelse(x == "y",T,F)) %>%
  arrange(Gene) %>%
  mutate(gene_symbol = get_standard_symbols(Gene),
         pan_cancer = Gene %in% ess.genes$pan_cancer_Sanger,
         bagel_essential = Gene %in% ess.genes$bagel_essential,
         bagel_nonessential = Gene %in% ess.genes$bagel_nonessential)


load(file="../data/ess.genes.rda")
yusa.v3 <- read.csv("../data/CRISPR_libraries/yusa_v3-suppl-table-S3.csv", stringsAsFactors = F) %>%
  left_join(read.table("../data/cleanr.tsv",header=T, stringsAsFactors = F), by = c("gRNA_ID" = "CODE"))

```

# Sanity checks on input genes

```{r}
cat("Pan-cancer Sanger set comparison:")
kable(table(genes$PanCancer_essential,genes$pan_cancer))

cat("Pan Cancer essentials in Hart's Neutral gene list:")
kable(
  genes %>%
    filter(PanCancer_essential == F & pan_cancer == T)
)

cat("Expected numbers of essential groups:")
sum(genes$Early_essential) == params$early
sum(genes$Mid_essential) == params$mid
sum(genes$Late_essential) == params$late

```

# Build benchmarking library

```{r,eval=F}
# We don't need to pull down base 20 for the Yusa library:
# Kosuke replaced base 20 with a G and the guides only match 19 bases of the genome.
yusa <- yusa.v3 %>%
  filter(GENES %in% genes$Gene) %>%
  # All guides are 19bp in the 'cleanr' file - base 20 is variable even though 21 and 22 are always GG. 
  # Hence, must extract base 20 so guides can be correctly synthesised.
  rowwise() %>%
  mutate(sequence = ifelse(STRAND == "+", as.character(Hsapiens[[paste("chr",CHRM,sep="")]][STARTpos:ENDpos]),
                           as.character(reverseComplement(Hsapiens[[paste("chr",CHRM,sep="")]][STARTpos:ENDpos])))) %>%
  ungroup()
```

```{r}
# Libraries.
yusa <- yusa.v3 %>%
  filter(!grepl("Control",gRNA_ID)) %>%
  arrange(Gene,gRNA_ID) %>%
  add_syms(gene_col = "Gene") %>%
  filter(gene_symbol %in% genes$gene_symbol) %>%
  mutate(library = "Yusa_v3", target_exon = NA) %>%
  rename(grna_id = gRNA_ID, gene_name.original = Gene, strand = STRAND, sequence = seq) %>%
  arrange(gene_symbol, grna_id) %>%
  select(library, gene_name.original, gene_symbol, grna_id, strand, target_exon, sequence)

brunello <- read.csv("../data/CRISPR_libraries/brunello_grnas.csv", stringsAsFactors = F) %>%
  arrange(Target.Gene.Symbol) %>%
  add_syms(gene_col = "Target.Gene.Symbol") %>%
  filter(gene_symbol %in% genes$gene_symbol) %>%
  mutate(library = "Brunello", num = 1:nrow(.), grna_id = paste("gattinara.",num,sep=""),
         Strand = ifelse(Strand == "sense","+","-")) %>%
  rename(gene_name.original = Target.Gene.Symbol, sequence = sgRNA.Target.Sequence,
         strand = Strand, target_exon = Exon.Number) %>%
  arrange(gene_symbol, grna_id) %>%
  select(library, gene_name.original, gene_symbol, grna_id, strand, target_exon, sequence)

toronto <- read.csv("../data/CRISPR_libraries/toronto_grnas.csv", stringsAsFactors = F) %>%
  arrange(GENE) %>%
  add_syms(gene_col = "GENE") %>%
  filter(gene_symbol %in% genes$gene_symbol) %>%
  rename(grna_id = GUIDE_ID, gene_name.original = GENE, sequence = SEQUENCE, target_exon = TARGET.EXON) %>%
  mutate(library = "Toronto",target_exon = as.integer(gsub("^exon_(\\d+)$","\\1",target_exon))) %>%
  rowwise() %>%
  mutate(strand = tail(unlist(strsplit(grna_id,"_")),1)) %>%
  ungroup() %>%
  arrange(gene_symbol, grna_id) %>%
  select(library, gene_name.original, gene_symbol, grna_id, strand, target_exon, sequence)

gecko <- read.csv("../data/CRISPR_libraries/gecko_grnas.csv", stringsAsFactors = F) %>%
  arrange(gene_id) %>%
  add_syms(gene_col = "gene_id") %>%
  filter(gene_symbol %in% genes$gene_symbol) %>%
  rename(grna_id = UID, gene_name.original = gene_id, sequence = seq) %>%
  mutate(library = "Gecko", target_exon = NA, strand = NA) %>%
  arrange(gene_symbol, grna_id) %>%
  select(library, gene_name.original, gene_symbol, grna_id, strand, target_exon, sequence)

gattinara <- read.csv("../data/CRISPR_libraries/gattinara_grnas.csv", stringsAsFactors = F) %>%
  arrange(Gene.Symbol) %>%
  add_syms(gene_col = "Gene.Symbol") %>%
  filter(gene_symbol %in% genes$gene_symbol) %>%
  mutate(library = "Gattinara", target_exon = NA, num = 1:nrow(.),
         grna_id = paste("gattinara.",num,sep=""), target_exon = NA, strand = NA) %>%
  rename(gene_name.original = Gene.Symbol, sequence = Barcode.Sequence) %>%
  arrange(gene_symbol, grna_id) %>%
  select(library, gene_name.original, gene_symbol, grna_id, strand, target_exon, sequence)

```

```{r}
# Horizon gRNA library must be kept locally.
horizon <- read.csv("~/data/az_cruk/horizon_grnas.csv", stringsAsFactors = F) %>%
  arrange(GeneSymbol) %>%
  add_syms(gene_col = "GeneSymbol") %>%
  filter(gene_symbol %in% genes$gene_symbol) %>%
  mutate(library = "Horizon", target_exon = NA, strand = substr(GenomicLocation,6,6)) %>%
  rename(gene_name.original = GeneSymbol, grna_id = CatalogNumber, sequence = TargetSequence) %>%
  arrange(gene_symbol, grna_id) %>%
  select(library, gene_name.original, gene_symbol, grna_id, strand, target_exon, sequence)

```

```{r}
bm_library <- rbind(yusa, brunello, toronto, gecko, gattinara, horizon) %>%
  # Overlapping guides.
  group_by(sequence) %>%
  mutate(status = ifelse(n() == 1,"Unique",paste(unique(library),collapse=".")),
         status = ifelse(grepl("\\.",status),status,"Unique")) %>%
  ungroup() %>%
  filter(!duplicated(sequence))

```

# Library stats

## Gene coverage

```{r}
bm_library %>%
  group_by(library) %>%
  summarise(number_genes = length(unique(gene_name.original))) %>%
  ungroup() %>%
  kable()

```


