---
title: "Benchmarking sgRNA libraries: Build one mixed library for reduced set of essentials and non-essentials"
author: "Alex Kalinka"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: hide
    depth: 3
    highlight: tango
    number_sections: true
    theme: spacelab
    toc: true
    toc_float: true
params:
  early: !r 86
  mid: !r 77
  late: !r 87
---

# Setup

```{r setup}
options(warn=-1)
suppressMessages(library(knitr))
suppressMessages(library(tidyr))
suppressMessages(library(magrittr))
suppressMessages(library(ggplot2))
suppressMessages(library(dplyr))
# Try to convert all genes into standard gene symbols.
suppressMessages(library(org.Hs.eg.db))
suppressMessages(library(biomaRt))
suppressMessages(library(Rbowtie2))


filter <- dplyr::filter
select <- dplyr::select
rename <- dplyr::rename
arrange <- dplyr::arrange

ensembl <- useMart("ensembl",dataset="hsapiens_gene_ensembl")


get_standard_symbols <- function(gene_names){
  syms <- AnnotationDbi::select(org.Hs.eg.db, gene_names, "SYMBOL", "ALIAS") %>%
    rowwise() %>%
    mutate(SYMBOL = ifelse(is.na(SYMBOL),ALIAS,SYMBOL)) %>%
    ungroup() %>%
    # Resolve many:1 mappings.
    group_by(ALIAS) %>%
    mutate(SYMBOL = ifelse(n() > 1, ALIAS, SYMBOL)) %>%
    filter(!duplicated(ALIAS)) %>%
    ungroup()
  return(syms$SYMBOL)
}

add_syms <- function(data, gene_col){
  reps <- data %>%
    group_by(!!sym(gene_col)) %>%
    summarise(num = n()) %>%
    ungroup()
  syms <- get_standard_symbols(unique(unlist(data[,gene_col])))
  if(length(syms) != length(unique(unlist(data[,gene_col]))))
    stop("unequal number of symbols returned")
  data %<>%
    mutate(gene_symbol = rep(syms, reps$num))
  return(data)
}


# Functions to:
# 1. Get missing gene coding regions.
# 2. Build a FASTA file of the missing genes.
# 3. Build into a bowtie reference.
# 4. Short-read alignment of the gRNAs.
build_missing_gene_bowtie_ref <- function(genes, fasta_out, bowtie_ref_dir){
  # 1. Download coding regions from biomart.
  seqs <- getBM(attributes = c("hgnc_symbol","coding"), filters="hgnc_symbol", values=genes, mart=ensembl) %>%
    filter(!is.na(coding)) %>%
    rowwise() %>%
    mutate(len = nchar(coding)) %>%
    ungroup() %>%
    # 2. Select longest seq for each gene.
    group_by(hgnc_symbol) %>%
    filter(len == max(len)) %>%
    filter(n()==1) %>%
    ungroup()
  seq_inp <- seqs$coding
  names(seq_inp) <- seqs$hgnc_symbol
  seqs <- DNAStringSet(seq_inp)
  # 3. BUild FASTA.
  writeXStringSet(seqs, fasta_out, format = "fasta")
  # 4. Make bowtie reference.
  bowtie2_build(fasta_out, bt2Index=bowtie_ref_dir)
  return(seqs)
}


```


```{r}
load(file="../data/ess.genes.rda")

genes <- read.csv("../data/benchmarking_genes.csv", stringsAsFactors = F) %>%
  mutate_at(vars(-Gene), .funs = function(x) ifelse(x == "y",T,F)) %>%
  arrange(Gene) %>%
  mutate(gene_symbol = get_standard_symbols(Gene),
         pan_cancer = Gene %in% ess.genes$pan_cancer_Sanger,
         bagel_essential = Gene %in% ess.genes$bagel_essential,
         bagel_nonessential = Gene %in% ess.genes$bagel_nonessential)


load(file="../data/ess.genes.rda")
yusa.v3 <- read.csv("../data/CRISPR_libraries/yusa_v3-suppl-table-S3.csv", stringsAsFactors = F) %>%
  left_join(read.table("../data/cleanr.tsv",header=T, stringsAsFactors = F), by = c("gRNA_ID" = "CODE"))

```

# Sanity checks on input genes

```{r}
cat("Pan-cancer Sanger set comparison:")
kable(table(genes$PanCancer_essential,genes$pan_cancer))

cat("Pan Cancer essentials in Hart's Neutral gene list:")
kable(
  genes %>%
    filter(PanCancer_essential == F & pan_cancer == T)
)

cat("Expected numbers of essential groups:")
cat("Early:")
sum(genes$Early_essential) == params$early
cat("Mid:")
sum(genes$Mid_essential) == params$mid
cat("Late:")
sum(genes$Late_essential) == params$late

```

# Benchmarking library

## Build

### Standard

```{r}
# Libraries.
yusa <- yusa.v3 %>%
  filter(!grepl("Control",gRNA_ID)) %>%
  arrange(Gene,gRNA_ID) %>%
  add_syms(gene_col = "Gene") %>%
  filter(gene_symbol %in% genes$gene_symbol) %>%
  mutate(library = "Yusa_v3", target_exon = NA, 
         gene_symbol.fgc = genes$Gene[match(gene_symbol,genes$gene_symbol)]) %>%
  rename(grna_id = gRNA_ID, gene_name.original = Gene, strand = STRAND, sequence = Guide_sequence) %>%
  arrange(gene_symbol, grna_id) %>%
  select(library, gene_name.original, gene_symbol.fgc, gene_symbol, grna_id, strand, target_exon, sequence)

brunello <- read.csv("../data/CRISPR_libraries/brunello_grnas.csv", stringsAsFactors = F) %>%
  arrange(Target.Gene.Symbol) %>%
  add_syms(gene_col = "Target.Gene.Symbol") %>%
  filter(gene_symbol %in% genes$gene_symbol) %>%
  mutate(library = "Brunello", num = 1:nrow(.), grna_id = paste("gattinara.",num,sep=""),
         Strand = ifelse(Strand == "sense","+","-"),
         gene_symbol.fgc = genes$Gene[match(gene_symbol,genes$gene_symbol)]) %>%
  rename(gene_name.original = Target.Gene.Symbol, sequence = sgRNA.Target.Sequence,
         strand = Strand, target_exon = Exon.Number) %>%
  arrange(gene_symbol, grna_id) %>%
  select(library, gene_name.original, gene_symbol.fgc, gene_symbol, grna_id, strand, target_exon, sequence)

toronto <- read.csv("../data/CRISPR_libraries/toronto_grnas.csv", stringsAsFactors = F) %>%
  arrange(GENE) %>%
  add_syms(gene_col = "GENE") %>%
  filter(gene_symbol %in% genes$gene_symbol) %>%
  rename(grna_id = GUIDE_ID, gene_name.original = GENE, sequence = SEQUENCE, target_exon = TARGET.EXON) %>%
  mutate(library = "Toronto",target_exon = as.integer(gsub("^exon_(\\d+)$","\\1",target_exon)),
         gene_symbol.fgc = genes$Gene[match(gene_symbol,genes$gene_symbol)]) %>%
  rowwise() %>%
  mutate(strand = tail(unlist(strsplit(grna_id,"_")),1)) %>%
  ungroup() %>%
  arrange(gene_symbol, grna_id) %>%
  select(library, gene_name.original, gene_symbol.fgc, gene_symbol, grna_id, strand, target_exon, sequence)

gecko <- read.csv("../data/CRISPR_libraries/gecko_grnas.csv", stringsAsFactors = F) %>%
  arrange(gene_id) %>%
  add_syms(gene_col = "gene_id") %>%
  filter(gene_symbol %in% genes$gene_symbol) %>%
  rename(grna_id = UID, gene_name.original = gene_id, sequence = seq) %>%
  mutate(library = "Gecko", target_exon = NA, strand = NA,
         gene_symbol.fgc = genes$Gene[match(gene_symbol,genes$gene_symbol)]) %>%
  arrange(gene_symbol, grna_id) %>%
  select(library, gene_name.original, gene_symbol.fgc, gene_symbol, grna_id, strand, target_exon, sequence)

gattinara <- read.csv("../data/CRISPR_libraries/gattinara_grnas.csv", stringsAsFactors = F) %>%
  arrange(Gene.Symbol) %>%
  add_syms(gene_col = "Gene.Symbol") %>%
  filter(gene_symbol %in% genes$gene_symbol) %>%
  mutate(library = "Gattinara", target_exon = NA, num = 1:nrow(.),
         grna_id = paste("gattinara.",num,sep=""), target_exon = NA, strand = NA,
         gene_symbol.fgc = genes$Gene[match(gene_symbol,genes$gene_symbol)]) %>%
  rename(gene_name.original = Gene.Symbol, sequence = Barcode.Sequence) %>%
  arrange(gene_symbol, grna_id) %>%
  select(library, gene_name.original, gene_symbol.fgc, gene_symbol, grna_id, strand, target_exon, sequence)

```

### Horizon

```{r}
# Horizon gRNA library must be kept locally.
horizon <- read.csv("~/data/az_cruk/horizon_grnas.csv", stringsAsFactors = F) %>%
  arrange(GeneSymbol) %>%
  add_syms(gene_col = "GeneSymbol") %>%
  filter(gene_symbol %in% genes$gene_symbol) %>%
  mutate(library = "Horizon", target_exon = NA, strand = substr(GenomicLocation,6,6),
         gene_symbol.fgc = genes$Gene[match(gene_symbol,genes$gene_symbol)]) %>%
  rename(gene_name.original = GeneSymbol, grna_id = CatalogNumber, sequence = TargetSequence) %>%
  arrange(gene_symbol, grna_id) %>%
  select(library, gene_name.original, gene_symbol.fgc, gene_symbol, grna_id, strand, target_exon, sequence)

```

### Croatan

```{r}
croatan.all <- read.table("../data/CRISPR_libraries/HG19_Library_For_Build.with_gene.txt",
                      header=F, stringsAsFactors = F)
croatan <- croatan.all %>%
  select(-V3,-V4,-V5) %>%
  rowwise() %>%
  mutate(gRNA_sequence.1 = unlist(strsplit(V2,"__"))[2],
         gRNA_sequence.2 = unlist(strsplit(V2,"__"))[6]) %>%
  ungroup() %>%
  select(-V2) %>%
  gather(key = "guide_number", value = "context_sequence", -V1) %>%
  rename(gene = V1) %>%
  arrange(gene, guide_number, context_sequence) %>%
  rowwise() %>%
  mutate(sequence = substr(context_sequence,5,24), PAM = substr(context_sequence,25,27),) %>%
  ungroup() %>%
  arrange(gene) %>%
  add_syms(gene_col = "gene") %>%
  right_join(genes, by = "gene_symbol") %>%
  #filter(gene_symbol %in% genes$gene_symbol) %>%
  mutate(library = "Croatan", target_exon = NA, strand = NA,
         gene_symbol.fgc = genes$Gene[match(gene_symbol,genes$gene_symbol)],
         num = 1:nrow(.), grna_id = paste("Croatan.",num,sep="")) %>%
  rename(gene_name.original = gene) %>%
  arrange(gene_symbol, grna_id) %>%
  select(library, gene_name.original, gene_symbol.fgc, gene_symbol, grna_id, strand, target_exon, sequence)

```

### Final

```{r}
bm_library <- rbind(yusa, brunello, toronto, gecko, gattinara, horizon) %>%
  # Overlapping guides.
  # Must group by 19bp guides so we can include Yusa.
  rowwise() %>%
  mutate(seq_19 = ifelse(library!="Yusa_v3",substr(sequence,2,20),sequence)) %>%
  ungroup() %>%
  group_by(seq_19) %>%
  mutate(library = ifelse(n() == 1, library, paste(unique(library),collapse="."))) %>%
  ungroup() %>%
  filter(!duplicated(sequence))

```

## Find missing genes using gRNA sequences

```{r}


```


# Library stats

## Gene coverage

```{r}
# Summarised by library.
bm_library %>%
  

# Total breakdown.
bm_library %>%
  group_by(library) %>%
  summarise(number_genes = length(unique(gene_symbol.fgc))) %>%
  ungroup() %>%
  arrange(desc(number_genes)) %>%
  kable()

```


